input_explainer.md

```diff
@@ -0,0 +1,406 @@
+
+TODO: Rationalize with onvrdisplaypointerrestricted from WebVR 1.1.
+
+## Appendix A: I don’t understand why this is a new API. Why can’t we use…
+
+### Pointer events
+Pointer events and their various sub-parts (mouse, touch, etc.) were created with the needs of a 2D web in mind, and serve that purpose well. Trying to bolt on an understanding of 3D space would both over-complicate the APIs with functionality that most applications don't need and make VR/AR input a second-class citizen within the larger pointer even model.
+
+We do expect that many future VR/AR uses (such as interacting with 2D DOM rectangles in 3D space) will need to emulate 2D input methods, producing appropriately transformed pointer events in response to VR/AR input.
+
+### The Gamepad API
+This was actually the first thing that we tried, extending the API with the necessary pose data. It proved to be mechanically feasible, but provided a confusing, non-semantic view of VR/AR input that left developers guessing as to what each button meant based on the device name string. Obviously that's a fairly fragile model, and not one that can reasonably extend to a robust ecosystem of hundreds of different devices. It also lacks many of the higher level concepts that we feel developers will eventually need (events for user initiation, ray generation, action mapping across devices, controller visualization, etc.)
+
+## Appendix B: Proposed IDL
+
+```webidl
+//
+// Controllers
+//
+
+interface VRInputSource {
+};
+
+interface VRGamepadInputSource : VRInputSource {
+  readonly attribute Gamepad gamepad;
+};
+
+enum VRHandedness {
+  "",
+  "left",
+  "right"
+};
+
+interface VRController : VRInputSource {
+  readonly attribute VRHandedness handedness;
+
+  // TODO: Better names for these two properties.
+  readonly attribute boolean supportsPointing; 
+  readonly attribute boolean supportsGrabbing; 
+
+  readonly attribute VRControllerElementMap elements;
+  readonly attribute VRControllerMesh mesh;
+};
+
+interface VRControllerElementMap {
+  readonly maplike<DOMString, VRControllerElement>;
+};
+
+interface VRControllerElement {
+  readonly attribute boolean pressed;
+  readonly attribute boolean touched;
+  readonly attribute double  value;
+  readonly attribute double? xAxis;
+  readonly attribute double? yAxis;
+};
+
+interface VRControllerMesh {
+  Promise<ArrayBuffer> requestBuffer();
+};
+
+//
+// Frame
+//
+
+interface VRInputPose {
+  readonly attribute Float32Array? gripPoseMatrix;
+  readonly attribute Float32Array? pointerPoseMatrix;
+  // velocity/acceleration here? v0.1+?
+  // readonly attribute VRHandPose? handPose; // v0.1+?
+};
+
+partial interface VRPresentationFrame {
+  VRInputPose? getInputPose(VRInputSource inputSource, VRCoordinateSystem coordinateSystem);
+};
+
+//
+// Session
+//
+
+partial interface VRSession {
+  attribute EventHandler onselect;
+
+  attribute EventHandler oncontrolleradded;
```

Comment:
maybe `oncontrollerconnect` \ `oncontrollerdisconnect` for consistency with `ondeviceconnect` \ `ondevicedisconeect` of the WebVR 2.0 API (https://w3c.github.io/webvr/spec/latest/#dom-vr-ondeviceconnect)

Reply:
This seems face-palm-worthy obvious now that you've pointed it out. 😅 I'm all for the change.

I do seem to recall that @NellWaliczek and friends may have preferred the "added" verbiage simply because it may apply better to inputs like hand tracking where hands may fall in and out of tracking but not be "connecting" and "disconnecting" in the traditional sense. I feel like the terminology isn't a big deal and I'd prefer consistency, but I am curious to hear the Microsoft team's opinion.